src/util/address.rs

```diff
@@ -143,11 +154,167 @@ impl FromStr for AddressType {
+    pub fn from_u5(value: ::bech32::u5) -> Result<Self, Error> {
+        WitnessVersion::from_ver_u8(value.to_u8())
+    }
+
+    /// Takes byte value and returns either corresponding version of
+    /// the Witness program (for opcodes in range of `OP_0`..`OP_16`) or
+    /// [`WitnessVersionError`] error for the rest of opcodes
+    pub fn from_ver_u8(value: u8) -> Result<Self, Error> {
+        Ok(match value {
+            0 => WitnessVersion::V0,
+            1 => WitnessVersion::V1,
+            2 => WitnessVersion::V2,
+            3 => WitnessVersion::V3,
+            4 => WitnessVersion::V4,
+            5 => WitnessVersion::V5,
+            6 => WitnessVersion::V6,
+            7 => WitnessVersion::V7,
+            8 => WitnessVersion::V8,
+            9 => WitnessVersion::V9,
+            10 => WitnessVersion::V10,
+            11 => WitnessVersion::V11,
+            12 => WitnessVersion::V12,
+            13 => WitnessVersion::V13,
+            14 => WitnessVersion::V14,
+            15 => WitnessVersion::V15,
+            16 => WitnessVersion::V16,
+            wrong => Err(Error::InvalidWitnessVersion(wrong))?,
+        })
+    }
+
+    /// Takes bitcoin Script opcode and returns either corresponding version of
+    /// the Witness program (for opcodes in range of `OP_0`..`OP_16`) or
+    /// [`WitnessVersionError`] error for the rest of opcodes
+    pub fn from_opcode(opcode: opcodes::All) -> Result<Self, Error> {
+        match opcode.into_u8() {
+            0 => Ok(WitnessVersion::V0),
+            ver if ver >= opcodes::all::OP_PUSHNUM_1.into_u8() && ver <= opcodes::all::OP_PUSHNUM_16.into_u8() =>
+                WitnessVersion::from_ver_u8(ver - opcodes::all::OP_PUSHNUM_1.into_u8() + 1),
+            _ => Err(Error::UnparsableWitnessVersion)
+        }
+    }
+
+    /// Takes bitcoin Script instruction (parsed opcode) and returns either
+    /// corresponding version of the Witness program (for push-num instructions)
+    /// or [`WitnessVersionError`] error for the rest of opcodes
+    pub fn from_instruction(instruction: Instruction) -> Result<Self, Error> {
+        match instruction {
+            Instruction::Op(op) => WitnessVersion::from_opcode(op),
+            Instruction::PushBytes(bytes) if bytes.len() == 0 => Ok(WitnessVersion::V0),
+            Instruction::PushBytes(_) => Err(Error::UnparsableWitnessVersion),
+        }
+    }
+
+    /// Returns integer version number representation for a given [`WitnessVersion`] value.
+    /// NB: this is not the same as an integer representation of the opcode signifying witness
+    /// version in bitcoin script. Thus, there is no function to directly convert witness version
+    /// into a byte since the conversion requires context (bitcoin script or just a version number)
+    pub fn into_ver_u8(self) -> u8 {
+        self as u8
+    }
+
+    /// Determine the checksum variant. See BIP-0350 for specification.
+    pub fn bech32_variant(&self) -> bech32::Variant {
```

Comment:
feel free to ignore since it's a late comment but I was wondering if it has been considered as `impl Into<bech32::Variant> for WitnessVersion` ?

Reply:
I guess `From` rather than `Into` but it seems to make sense.

Reply:
yeah, sorry I meant `impl From<WitnessVersion> for bech32::Variant`

Reply:
Added in 747d97cdd6e2d66fa10a8c0cf6bac2e1d06dadde